Error when cancelling editing of a specific section

1. On page: strictdoc/docs/strictdoc_01_user_guide.html
2. Go to: **2.1. Hello World** (The error appears specifically on this section. There is no error in the neighboring sections)

3. edit:
```
INFO:     127.0.0.1:51604 - "GET /actions/document/edit_section/78139f6f935c4964a36b84ccb0dd4531 HTTP/1.1" 200 OK
```
4. cancel:
```
INFO:     127.0.0.1:51682 - "GET /actions/document/cancel_edit_section?section_mid=78139f6f935c4964a36b84ccb0dd4531 HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/fastapi/applications.py", line 271, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/applications.py", line 118, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/middleware/cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/routing.py", line 706, in __call__
    await route.handle(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/fastapi/routing.py", line 237, in app
    raw_response = await run_endpoint_function(
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/fastapi/routing.py", line 165, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/Users/mettta/Projects/dev/strictdoc/./strictdoc/server/routers/main_router.py", line 595, in cancel_edit_section
    output = template.render(
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/Users/mettta/Projects/dev/strictdoc/.venv-default/lib/python3.9/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/Users/mettta/Projects/dev/strictdoc/strictdoc/export/html/templates/actions/document/edit_section/stream_updated_section.jinja.html", line 3, in top-level template code
    {% include "_shared/section.jinja.html" %}
  File "/Users/mettta/Projects/dev/strictdoc/strictdoc/export/html/templates/_shared/section.jinja.html", line 17, in top-level template code
    {% include "_shared/requirement_block/free_text.jinja.html" %}
  File "/Users/mettta/Projects/dev/strictdoc/strictdoc/export/html/templates/_shared/requirement_block/free_text.jinja.html", line 4, in top-level template code
    {{ renderer.render_free_text(document_type, free_text) }}
  File "/Users/mettta/Projects/dev/strictdoc/./strictdoc/export/html/renderers/markup_renderer.py", line 99, in render_free_text
    href = self.link_renderer.render_requirement_link(
  File "/Users/mettta/Projects/dev/strictdoc/./strictdoc/export/html/renderers/link_renderer.py", line 71, in render_requirement_link
    assert isinstance(context_document, Document), context_document
AssertionError: Section(uid = NoneType(...), custom_level = "", title = "Examples", free_texts = [], section_contents = [3 elements], _tx_position = 2271, _tx_position_end = 4408, parent = Document(...), reserved_uid = NoneType(...), ng_resolved_custom_level = "", ng_level = 1, ng_has_requirements = False, ng_document_reference = DocumentReference(...), context = SectionContext(...), node_id = "5f77c9345733421f94f7a754c0dda164")
```

--- a/strictdoc/export/html/renderers/markup_renderer.py
+++ b/strictdoc/export/html/renderers/markup_renderer.py
@@ -25,8 +25,6 @@ class MarkupRenderer:
         link_renderer: LinkRenderer,
         context_document: Optional[Document],
     ) -> "MarkupRenderer":
-        assert isinstance(traceability_index, TraceabilityIndex)
-        assert isinstance(link_renderer, LinkRenderer)
         html_fragment_writer: Union[
             Type[RstToHtmlFragmentWriter],
             Type[HTMLFragmentWriter],
@@ -52,6 +50,12 @@ class MarkupRenderer:
         link_renderer: LinkRenderer,
         context_document: Optional[Document],
     ):
+        assert isinstance(traceability_index, TraceabilityIndex)
+        assert isinstance(link_renderer, LinkRenderer)
+        assert context_document is None or isinstance(
+            context_document, Document
+        ), context_document
+
         self.fragment_writer = fragment_writer
         self.traceability_index = traceability_index
         self.link_renderer: LinkRenderer = link_renderer

--- a/strictdoc/server/routers/main_router.py
+++ b/strictdoc/server/routers/main_router.py
@@ -598,7 +598,7 @@ def create_main_router(
             markup="RST",
             traceability_index=export_action.traceability_index,
             link_renderer=link_renderer,
-            context_document=section.parent,
+            context_document=section.document,
         )
         output = template.render(
             renderer=markup_renderer,

--- a/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/expected_output/document.sdoc
+++ b/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/expected_output/document.sdoc
@@ -1,0 +1,24 @@
+[DOCUMENT]
+TITLE: Document 1
+
+[FREETEXT]
+Hello world!
+[/FREETEXT]
+
+[SECTION]
+TITLE: Section 1
+
+[FREETEXT]
+This is a free text of the section 1.
+[/FREETEXT]
+
+[SECTION]
+TITLE: Section 1.1
+
+[FREETEXT]
+This is a free text of the section 1.1.
+[/FREETEXT]
+
+[/SECTION]
+
+[/SECTION]

--- a/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/input/document.sdoc
+++ b/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/input/document.sdoc
@@ -1,0 +1,24 @@
+[DOCUMENT]
+TITLE: Document 1
+
+[FREETEXT]
+Hello world!
+[/FREETEXT]
+
+[SECTION]
+TITLE: Section 1
+
+[FREETEXT]
+This is a free text of the section 1.
+[/FREETEXT]
+
+[SECTION]
+TITLE: Section 1.1
+
+[FREETEXT]
+This is a free text of the section 1.1.
+[/FREETEXT]
+
+[/SECTION]
+
+[/SECTION]

--- a/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/test_UC08_T05_cancel_edit_nested_section.py
+++ b/tests/end2end/document/UC08_edit_section/UC08_T05_cancel_edit_nested_section/test_UC08_T05_cancel_edit_nested_section.py
@@ -1,0 +1,40 @@
+from seleniumbase import BaseCase
+
+from tests.end2end.end2end_test_setup import End2EndTestSetup
+from tests.end2end.helpers.constants import NODE_2
+from tests.end2end.helpers.screens.document.form_edit_section import (
+    Form_EditSection,
+)
+from tests.end2end.helpers.screens.document_tree.screen_document_tree import (
+    Screen_DocumentTree,
+)
+from tests.end2end.server import SDocTestServer
+
+
+class Test_UC08_T05_CancelEditNestedSection(BaseCase):
+    def test_01(self):
+        test_setup = End2EndTestSetup(path_to_test_file=__file__)
+
+        with SDocTestServer(
+            input_path=test_setup.path_to_sandbox
+        ) as test_server:
+            self.open(test_server.get_host_and_port())
+
+            screen_document_tree = Screen_DocumentTree(self)
+
+            screen_document_tree.assert_on_screen()
+            screen_document_tree.assert_contains_document("Document 1")
+
+            screen_document = screen_document_tree.do_click_on_first_document()
+
+            screen_document.assert_on_screen()
+            screen_document.assert_is_document_title("Document 1")
+
+            screen_document.assert_text("Hello world!")
+
+            form_edit_section: Form_EditSection = (
+                screen_document.do_open_form_edit_section(NODE_2)
+            )
+            form_edit_section.do_form_cancel()
+
+        assert test_setup.compare_sandbox_and_expected_output()

--- a/tests/end2end/helpers/constants.py
+++ b/tests/end2end/helpers/constants.py
@@ -17,6 +17,7 @@ NODE_0 = 1
 # which is actually the first node
 # (section or requirement) on the page.
 NODE_1 = 2
+NODE_2 = 3
 
 TEXT_WITH_TRAILING_WHITESPACES = """
 Hello world!    

